File: app/src/main/java/com/hjc/reader/viewmodel/wan/TagViewModel.java
package com.hjc.reader.viewmodel.wan;


import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.blankj.utilcode.util.ToastUtils;
import com.hjc.baselib.http.RxSchedulers;
import com.hjc.baselib.viewmodel.CommonViewModel;
import com.hjc.reader.bean.response.WanListBean;
import com.hjc.reader.http.RetrofitHelper;
import com.hjc.reader.http.observer.BasePageObserver;

import java.util.List;

public class TagViewModel extends CommonViewModel {

    // 知识体系tag下文章列表数据
    private MutableLiveData<List<WanListBean.DataBean.DatasBean>> articleLiveData = new MutableLiveData<>();

    public TagViewModel(@NonNull Application application) {
        super(application);
    }


    /**
     * 获取知识体系tag下文章列表数据
     *
     * @param isFirst 是否第一次加载
     */
    public void loadArticleList(int page, int id, boolean isFirst) {
        RetrofitHelper.getInstance().getWanAndroidService()
                .getWanList(page, id)
                .compose(RxSchedulers.ioToMain())
                .subscribe(new BasePageObserver<WanListBean>(this, isFirst) {
                    @Override
                    public void onSuccess(WanListBean result) {
                        if (result.getErrorCode() == 0) {
                            WanListBean.DataBean dataBean = result.getData();
                            if (dataBean != null) {
                                List<WanListBean.DataBean.DatasBean> dataList = dataBean.getDatas();
                                if (dataList != null && dataList.size() > 0) {
                                    articleLiveData.setValue(dataList);
                                } else {
                                    articleLiveData.setValue(null);
                                    if (page == 0) {
                                        ToastUtils.showShort("暂无数据");
                                    } else {
                                        ToastUtils.showShort("没有更多数据了");
                                    }
                                }
                            }
                        } else {
                            ToastUtils.showShort(result.getErrorMsg());
                        }
                    }
                });
    }

    public MutableLiveData<List<WanListBean.DataBean.DatasBean>> getArticleLiveData() {
        return articleLiveData;
    }

}

File: module-wan/src/main/java/com/hjc/learn/wan/viewmodel/TagViewModel.kt
package com.hjc.learn.wan.viewmodel

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.blankj.utilcode.util.ToastUtils
import com.hjc.learn.wan.model.WanModel
import com.hjc.library_common.viewmodel.KotlinViewModel
import com.hjc.library_net.entity.WanArticleBean
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/**
 * @Author: HJC
 * @Date: 2021/2/23 9:50
 * @Description: 知识体系标签ViewModel
 */
class TagViewModel(application: Application) : KotlinViewModel(application) {

    private val mModel = WanModel()

    // 知识体系tag下文章列表数据
    val articleData = MutableLiveData<MutableList<WanArticleBean>>()

    // 刷新数据
    val refreshData = MutableLiveData<Boolean>()

    /**
     * 获取知识体系tag下文章列表数据
     *
     * @param isFirst 是否第一次加载
     */
    fun loadArticleList(page: Int, id: Int, isFirst: Boolean) {
        launchWan({ mModel.getWanList(page, id) }, { result ->
            refreshData.value = true

            val data = result?.datas
            data?.let {
                if (it.size > 0) {
                    showContent()
                    articleData.value = it
                } else {
                    if (page == 0) {
                        showEmpty()
                    } else {
                        showContent()
                        ToastUtils.showShort("没有更多数据了")
                    }
                }
            }
        }, { e ->
            refreshData.value = true

            if (e is UnknownHostException || e is SocketTimeoutException) {
                showError()
            } else {
                showTimeout()
            }
        }, isShowProgress = isFirst)
    }
}